lib/redis.js

```diff
@@ -1,79 +1,135 @@
-      settings = Object.assign({}, settings);
+      return client;
+    } else {
+      isNewClient = false;
     }
+  } else {
+    client = Redis.createClient(settings);
+  }
 
-    client = redis.createClient(settings);
+  if (getClientID && isNewClient) {
+    // Best-effort attempt to get the client ID. Might fail due to connection
+    // errors or due to an incompatible redis version.
+    client.client('ID').then(
+      (id) => clientIDs.set(client, id),
+      () => {}
+    );
   }
 
-  // Wait for the client to be ready, then resolve with the client itself.
-  return helpers.waitOn(client, 'ready', true).then(
-    () => client,
+  // Make sure the client is "ready" (connected etc) before we return it
+  try {
+    await helpers.waitOn(client, 'ready', true);
+  } catch (err) {
     // If we receive an error before the client becomes ready, we won't retain a
     // reference to it, so we should disconnect the client to prevent uncaught
     // exceptions.
-    (err) => (disconnect(client), Promise.reject(err))
-  );
+    if (isNewClient) {
+      disconnect(client);
+    }
+    throw err;
+  }
+  return client;
 }
 
-function disconnect(client) {
-  // Redis#end is deprecated for ioredis.
-  /* istanbul ignore if: this is only for ioredis */
-  if (client.disconnect) {
-    client.disconnect();
-  } else {
-    // true indicates that it should invoke all pending callbacks with an
-    // AbortError; we need this behavior.
-    client.end(true);
+async function cleanDisconnect(client, commandClient) {
+  if (!clientIDs.has(client)) {
+    // Disable auto-reconnection.
+    client.options.retryStrategy = false;
+    try {
+      await commandClient.client(
+        'KILL',
+        'ID',
+        String(clientIDs.get(client)),
+        'SKIPME',
+        'no'
+      );
+      await helpers.waitOn(client, 'end', true);
+      clientIDs.delete(client);
+      return;
+    } catch (err) {
+      // Fallthrough.
+    }
   }
+  disconnect(client);
+}
+
+function disconnect(client) {
+  // HACK: flushQueue is, according to ioredis, private. This is unfortunate,
+  // because we need this in order to properly clean up during brpoplpush
+  // commands.
+  client.flushQueue(new AbortError());
+  client.disconnect();
 }
 
 function isAbortError(err) {
-  // node_redis has a designated class for abort errors, but ioredis just has
-  // a constant message defined in a utils file.
-  return (
-    err.name === 'AbortError' ||
-    /* istanbul ignore next: this is only for ioredis */
-    err.message === 'Connection is closed.'
-  );
+  // Checking for a particular constant message defined in ioredis's utils
+  // TODO: can we use some other field?
+  return err.name === 'AbortError' || err.message === 'Connection is closed.';
+}
+
+const rBoundary = /\b/g;
+function errorCode(err) {
+  rBoundary.lastIndex = 1;
+  const { message } = err,
+    match = rBoundary.exec(err.message);
+  return err.message.slice(0, match ? match.index : undefined);
 }
```

Comment:
Probably worth submitting upstream, as this is a nice feature of NodeRedis and impacts #316